### PR TITLE
Feature/pool 182 redeemticketsinstantlyfrom exit fee

### PR DESCRIPTION
**Pull request by robsecord**

### CODE REVIEW

File: test/PeriodicPrizePool.test.js [hunk 1]
```diff
@@ -136,11 +136,54 @@ describe('PeriodicPrizePool contract', function() {
...
+      await ticket.mock.balanceOf.withArgs(wallet._address).returns(amount)
+      await cToken.mock.redeemUnderlying.withArgs(amount).returns(amount)
+      await token.mock.transfer.withArgs(wallet._address, amount).returns(true)
+      await ticket.mock.controllerBurn.withArgs(wallet2._address, wallet._address, amount).returns()
+      await ticketCredit.mock.controllerMint.withArgs(wallet._address, toWei('0')).returns()
+
+      // Test redeemTicketsInstantlyFrom
+      await expect(prizePool.connect(wallet2).redeemTicketsInstantlyFrom(wallet._address, amount, sponsoredExitFee))
+        .to.emit(prizePool, 'TicketsRedeemedInstantly')
+        .withArgs(wallet2._address, wallet._address, amount, toWei('0'), toWei('0'))
```
Comment (asselstine) [inline test/PeriodicPrizePool.test.js:186]: Make sure to test boundary conditions:

- partial sponsorship
- zero sponsorship 
- full sponsorship

File: contracts/periodic-prize-pool/PeriodicPrizePool.sol [hunk 2]
```diff
@@ -332,32 +333,41 @@ abstract contract PeriodicPrizePool is Timelock, BaseRelayRecipient, ReentrancyG
 
   function redeemTicketsInstantlyFrom(
     address from,
-    uint256 tickets
-  ) 
-    external nonReentrant returns (uint256) 
+    uint256 tickets,
+    uint256 prepaidExitFee
```
Comment (asselstine) [inline contracts/periodic-prize-pool/PeriodicPrizePool.sol:337]: Since it's called `sponsoredExitFee` in the event, it should be named the same here in the params.

File: contracts/periodic-prize-pool/PeriodicPrizePool.sol [hunk 3]
```diff
@@ -332,32 +333,41 @@ abstract contract PeriodicPrizePool is Timelock, BaseRelayRecipient, ReentrancyG
...
     uint256 userInterestRatioMantissa = _ticketInterestRatioMantissa(from);
-    uint256 exitFee = calculateExitFee(tickets, userInterestRatioMantissa);
+    uint256 totalExitFee = calculateExitFee(tickets, userInterestRatioMantissa);
 
-    // transfer the fee to this contract
-    _token().transferFrom(operator, address(this), exitFee);
+    uint256 userExitFee = totalExitFee;
+    uint256 sponsoredExitFee = (totalExitFee > prepaidExitFee) ? prepaidExitFee : totalExitFee;
+    if (totalExitFee > 0) {
+      userExitFee = (totalExitFee > prepaidExitFee) ? totalExitFee.sub(prepaidExitFee) : 0;
```
Comment (asselstine) [inline contracts/periodic-prize-pool/PeriodicPrizePool.sol:350]: This is a little confusing, as it's running the same condition `totalExitFee > prepaidExitFee` twice.

Since `sponsoredExitFee` represents the portion of the fee that is covered by the sponsor, it seems to me you can change lines `347`-`351` to:

```solidity
uint256 sponsoredExitFee = (totalExitFee > prepaidExitFee) ? prepaidExitFee : totalExitFee;
uint256 userExitFee = totalExitFee.sub(sponsoredExitFee);
```

Comment (robsecord) [reply contracts/periodic-prize-pool/PeriodicPrizePool.sol:350]: if totalExitFee is 0 we will get an underflow on `sub()`
